src/vanilla/store.ts

```diff
@@ -725,43 +710,89 @@ const buildStore = (
     sub: subscribeAtom,
     unstable_derive,
   }
-  if (import.meta.env?.MODE !== 'production') {
-    const devStore: DevStoreRev4 = {
-      // store dev methods (these are tentative and subject to change without notice)
-      dev4_get_internal_weak_map: () => ({
-        get: (atom) => {
-          const atomState = getAtomState(atom)
-          if (atomState.n === 0) {
-            // for backward compatibility
-            return undefined
+  return store
+}
+
+const deriveDevStore = (store: Store): Store & DevStoreRev4 => {
+  const proxyAtomStateMap = new WeakMap()
+  const debugMountedAtoms = new Set<AnyAtom>()
+  let savedGetAtomState: StoreArgs[0]
+  const RESTORE_ATOM = Symbol()
+  const derivedStore = store.unstable_derive(
+    (getAtomState, atomRead, atomWrite, atomOnMount) => {
+      savedGetAtomState = getAtomState
+      return [
+        (atom) => {
+          let proxyAtomState = proxyAtomStateMap.get(atom)
+          if (!proxyAtomState) {
+            const atomState = getAtomState(atom)
+            proxyAtomState = new Proxy(atomState, {
+              set(target, prop, value) {
+                if (prop === 'm') {
+                  debugMountedAtoms.add(atom)
+                }
+                return Reflect.set(target, prop, value)
+              },
+              deleteProperty(target, prop) {
+                if (prop === 'm') {
+                  debugMountedAtoms.delete(atom)
+                }
+                return Reflect.deleteProperty(target, prop)
+              },
+            })
+            proxyAtomStateMap.set(atom, proxyAtomState)
           }
-          return atomState
+          return proxyAtomState
         },
-      }),
-      dev4_get_mounted_atoms: () => debugMountedAtoms,
-      dev4_restore_atoms: (values) => {
-        const batch = createBatch()
-        for (const [atom, value] of values) {
-          if (hasInitialValue(atom)) {
-            const atomState = getAtomState(atom)
-            const prevEpochNumber = atomState.n
-            setAtomStateValueOrPromise(atom, atomState, value)
-            mountDependencies(batch, atom, atomState)
-            if (prevEpochNumber !== atomState.n) {
-              registerBatchAtom(batch, atom, atomState)
-              recomputeDependents(batch, atom, atomState)
+        atomRead,
+        (atom, getter, setter, ...args) => {
+          if (RESTORE_ATOM in atom) {
+            const values = atom[RESTORE_ATOM] as Iterable<
+              readonly [AnyAtom, AnyValue]
+            >
+            for (const [atom, value] of values) {
+              if (hasInitialValue(atom)) {
+                setter(atom as never, value)
+              }
```

Comment:
I think this is still an issue. `atom(0, () => {})`, needs the `init` value, but its write function should not be called.

I think you will need to add createPending to unstable_derive. Then you can update the atomState and adjust batch.

Reply:
> I think this is still an issue.

Thanks for the catch. I didn't notice there was a test case missing. Added in #2873.